src/Microsoft.AspNetCore.Mvc.Core/Internal/FileStreamResultExecutor.cs

```diff
@@ -18,19 +19,36 @@ public FileStreamResultExecutor(ILoggerFactory loggerFactory)
 
         public Task ExecuteAsync(ActionContext context, FileStreamResult result)
         {
-            SetHeadersAndLog(context, result);
-            return WriteFileAsync(context, result);
+            long? fileLength = null;
+            if (result.FileStream.CanSeek)
+            {
+                fileLength = result.FileStream.Length;
+            }
+
+            var (range, rangeLength, serveBody) = SetHeadersAndLog(
+                context,
+                result,
+                fileLength,
+                result.LastModified,
+                result.EntityTag);
+
+            if (!serveBody)
+            {
+                return Task.CompletedTask;
+            }
+
+            return WriteFileAsync(context, result, range, rangeLength);
         }
 
-        private static async Task WriteFileAsync(ActionContext context, FileStreamResult result)
+        private static async Task WriteFileAsync(ActionContext context, FileStreamResult result, RangeItemHeaderValue range, long rangeLength)
         {
             var response = context.HttpContext.Response;
             var outputStream = response.Body;
-
-            using (result.FileStream)
+            if (range != null && rangeLength == 0)
             {
-                await result.FileStream.CopyToAsync(outputStream, BufferSize);
+                return;
             }
+            await WriteFileAsync(context.HttpContext, result.FileStream, range, rangeLength);
```

Comment:
The `WriteFileAsync` in the base executor is async. If not, it is unreliable when writing to the response body. The functional tests fail as a result. So this needs to stay async/await. @Tratcher